Add common Upbound issue and PR automation

negz opened the pull request:
https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/creating-a-default-community-health-file

This PR adds global issue and PR templates per the above link. If I follow correctly these templates will become the default for any repos within the upbound org that don't have their own templates configured. If this PR is merged I'd like to do a pass through upbound repos and delete repo specific templates if they're not an improvement over these common ones.

https://docs.github.com/en/actions/learn-github-actions/sharing-workflows-with-your-organization
https://github.com/marketplace/actions/global-workflows-support

This PR also adds a broadly useful GitHub action that can be used to create a tag. As it stands we might need to use the 'global workflows' action to enable this, since GitHub appears to only support workflow _templates_ natively.

I considered adding more checklists to the issue and PR templates included in this PR, but it seems like we could automate most of the things we'd prompt the issue or PR author to handle. For example:

* Linking issues to an epic using an 'Epic: upbound/example#23' style comment.
* Adding a size estimate label to issues and aggregating size at the sprint column and/or epic level.
* Adding a size assessment to PRs (e.g. size/s, size/xl).
* Automatically backporting PRs to release branches.

negz committed:
Add common Upbound issue and PR automation

https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/creating-a-default-community-health-file

This commit adds global issue and PR templates per the above link. If I follow
correctly these templates will become the default for any repos within the
upbound org that don't have their own templates configured. If this PR is merged
I'd like to do a pass through upbound repos and delete repo specific templates
if they're not an improvement over these common ones.

https://docs.github.com/en/actions/learn-github-actions/sharing-workflows-with-your-organization
https://github.com/marketplace/actions/global-workflows-support

This commit also adds a broadly useful GitHub action that can be used to create
a tag. As it stands we might need to use the 'global workflows' action to enable
this, since GitHub appears to only support workflow _templates_ natively.

I considered adding more checklists to the issue and PR templates included in
this PR, but it seems like we could automate most of the things we'd prompt the
issue or PR author to handle. For example:

* Linking issues to an epic using an 'Epic: upbound/example#23' style comment.
* Adding a size estimate label to issues and aggregating size at the sprint
  column and/or epic level.
* Adding a size assessment to PRs (e.g. size/s, size/xl).
* Automatically backporting PRs to release branches.

Signed-off-by: Nic Cope <negz@rk0n.org>

negz commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,35 @@
+<!--
```
> Issue templates and their configuration file must be in a folder called .github/ISSUE_TEMPLATE. All other supported files must be in the root of the repository. For more information, see "Creating new files."

The [linked docs](https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/creating-a-default-community-health-file) mention that _issue templates_ specifically must be in a .github dir. It seems strange to me that this would not apply to PR templates; we probably need to merge this and test to figure out exactly where each file must be.

negz commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,35 @@
+
+- [ ] Read and followed Upbound's [contribution process](https://git.io/fj2m9).
+- [ ] Run `make reviewable test` to ensure this PR is ready for review.
+- [ ] Added `backport-release-x.y` labels to auto-backport this PR.
```
We could use https://mergify.io/ to automate the creation and merging of backport PRs for any PR with this label, which would reduce some of the toil involved in releasing bug fixes. @jbw976 recommended this tool based on its use in the @rook project. See https://github.com/rook/rook/pull/7972 as an example.

Note that Mergify is $8/mo/user for orgs with private repos like @upbound. There appear to be similar Actions available for free, for example https://github.com/tibdex/backport.

negz replied:
After much trial and error I landed on an action that works well.

negz commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,35 @@
+I have:
+
+- [ ] Read and followed Upbound's [contribution process](https://git.io/fj2m9).
+- [ ] Run `make reviewable test` to ensure this PR is ready for review.
```
Relatively few of our repositories support the `make reviewable` command; I think it's time we add it to the upbound/build repo as a standard command.

negz commented on .github/ISSUE_TEMPLATE/bug_report.md:
```diff
@@ -0,0 +1,29 @@
+Please let us know what behaviour you expected and how Upbound diverged from
+that behaviour.
+-->
+
```
I considered adding a checklist for the issue templates, which would include:

* Adding a size estimate label to your issue.
* Updating the relevant epic to link to the new issue.

I feel like we could automate a lot of this though. Given that GitHub Actions are mostly written in TypeScript it might be worth us putting a day aside for those of us with TS and GitHub Actions experience to hack on this. I suspect we'd be able to make everyone's lives a lot easier with a day or so spent on building automation.

negz replied:
Specifically https://github.com/mathiasjakobsen/github-estimates-action and https://github.com/kiali/kiali-epic-action go a long way toward aggregating issue estimates and keeping epics linked to sub-tasks, but unfortunately neither appear to support an org wide model in which epics might be in a different repo and project boards might be at the org level.

negz committed:
Add label prompts for size and epic links

The idea is that these labels will prompt folks to replace the labels with a
link to an epic and a size estimate. I prefer labels here to a checklist in the
issue because not all contributors will have the context and access needed to
create labels and link to epics.

Signed-off-by: Nic Cope <negz@rk0n.org>

grantgumina commented on .github/ISSUE_TEMPLATE/feature_request.md:
```diff
@@ -0,0 +1,22 @@
+---
+name: Feature Request
+about: Help us make Upbound more useful
+labels: enhancement, needs-triage, needs-size-label, needs-epic-link
```
I think `needs-size-label` and `needs-epic-link` are redundant. I think triaging an issue should mean swagging it, and parenting it with the right epic.

negz replied:
I think distinct labels work as better specific prompts vs a generic “triage” process while we don’t have automation. I don’t feel strongly though.

@grantgumina I wonder if it makes sense to address this by renaming `needs-triage` to something more specific and actionable. Not sure on the wording but I'm thinking something along the lines of "needs to be assessed for a sprint". Maybe `needs-priority`?

negz replied:
I went with `needs-project`.

negz committed:
Move workflows into the templates dir

GitHub doesn't let you _create_ workflows at an org level, but does let you add
templates to make creating workflows a little easier.

Signed-off-by: Nic Cope <negz@rk0n.org>

negz committed:
Use needs-project rather than needs-triage

This hopefully better indicates the intent; i.e. that we need to get the issue
into a sprint (or other?) project board.

Signed-off-by: Nic Cope <negz@rk0n.org>

negz committed:
Add a known working backport action

This action will help with our release process; it can automatically backport a
PR to a release branch.

Signed-off-by: Nic Cope <negz@rk0n.org>

negz committed:
Add an action to help with story points

This little action lets folks write /points comments (like /points 3). When they
do GitHub will automatically add a 'points/3' label and remove the
'needs-points-label' label if it exists.

Signed-off-by: Nic Cope <negz@rk0n.org>

grantgumina commented on .github/ISSUE_TEMPLATE/feature_request.md:
```diff
@@ -0,0 +1,22 @@
+### How could Upbound help solve your problem?
+<!--
+Let us know how you think Upbound could help with your use case. 
+-->
```
@negz What if we added the following checklist items as well to help with our process:

- [ ] I have linked to a parent issue or feature on our roadmap (optional)

negz replied:
Yeah, that was the original plan. I had hoped to automate that process using GitHub Actions. Sadly based on my investigation there isn't much we can do about that when the epics are in different repos.
